src/audio/kpb.c

```diff
@@ -551,8 +551,7 @@ static void kpb_buffer_data(struct comp_data *kpb, struct comp_buffer *source,
 			 * in this buffer, copy what's available and continue
 			 * with next buffer.
 			 */
-			assert(!memcpy_s(buff->w_ptr, space_avail, read_ptr,
-					 space_avail));
+			memcpy(buff->w_ptr, read_ptr, space_avail);
```

Comment:
@mrajwa Could you give any specific example? memcpy_s doesn't support wrapping automatically, so you still shouldn't fail at any time.

Reply:
@tlauda as you know history buffer uses large continuous blocks from various memory pools. Let's take HPSRAM as an example here. We buffer data internally.... at some point we have a situation like:

write pointer (history buffer) -> 0xbe01f480, 
read pointer (source component buffer ) -> 0xbe01f980, 
size to copy -> 1280 bytes.

now quick reminder of memcpy_s check:
`if ((dest + dest_size >= src && dest + dest_size <= src + src_size)`

Calculating if condition:

dest + dest_size = 0xbe01f480 + 1280 = 0xbe01f980
src + src_size = 0xbe01f980 + 1280  = 0xbe01fe80

so finally substituting values into if condition:

`if (TRUE && TRUE) `

This corner case where source component buffer happen to be right next to the end of our history buffer but such is life :)

Reply:
Although I think instead of removing these asserts and "safe" memcpy I would just go and allocate a byte more memory to avid such narrow cases.

Reply:
@mrajwa It looks like a workaround and not a fix. From your example it's clearly the wrong condition for buffer overlapping. I would go with something like this:
`if ((dest > src && dest < src + src_size) || (src > dest && src < dest + dest_size))`

Reply:
@tlauda, you just dropped the equal conditions but they are actually fine. After all source and destination should not be the same...assuming any reasonable copy right?

Reply:
@mrajwa I've proposed a little bit different condition. Your destination is ending at 0xbe01f980 and your source is starting at the same address. They are not overlapping, so where do you see the problem?

Reply:
@tlauda it is the question if we want actions like this:
`memcpy_s(0xbe01f980, 0x10, 0xbe01f980, 0x10)`

to raise an error or not. With your proposal it will be treated as permitted copy. Although it will not cause any issue it is a bit weird.

Reply:
> @tlauda it is the question if we want actions like this:
> `memcpy_s(0xbe01f980, 0x10, 0xbe01f980, 0x10)`
> 
> to raise an error or not. With your proposal it will be treated as permitted copy. Although it will not cause any issue it is a bit weird.

Overlapping source and destination have to raise error. It's like that in standard. Example that you given are overlapping buffers so yea, it's error.

Reply:
@tlauda @lgirdwood BTW shouldn't we have `restrict` in memcpys, to have it like in standard and maybe to get a performance (at least xtensa likes it)?

Reply:
@mrajwa @lgirdwood this is the bug I mentioned to you in the email, let me submit a patch for it.

Reply:
let's keep this assert usage here.